bazaarvoice/emodb
# Migration service + double writing

## Github Issue #

None

## What Are We Doing Here?

This PR is an implementation of the migration needed for [#116](https://github.com/bazaarvoice/emodb/pull/116). Because you cannot change the primary key in C*, we must be able to migrate all of Emo's existing delta to new C* tables without any outages.

This migration consists of two major parts.

### Double Table Writing
During the migration process we must write all new deltas to both tables, the old <placement>_delta table and the new <placement>_delta_v2 table. Even though we are writing to both tables, we will still only read from the old, as the new does not yet have all of the data.

### Move All Existing Deltas From Old Table to New Table
In order to do this, I have implemented a Emo migrator service mode. This service mode highly resembles Stash, as it operates by getting splits of 1,000,000 records, then copying them into the new table. Luckily, we do not have to worry about data inconsistencies because all deltas are immutable. Unlike Stash, this service only triggers migrations via API calls, rather than using a schedule. In order to ensure that all data is migrated, each existing placement needs to be migrated.

## How to Test and Verify

1. Check out Emo master
2. Build and run localy
3. Write a large amount of deltas
4. Copy the web-local/target/cassandra folder outside the Emodb repo
5. Check out this PR
6. Build Emo
7. Copy the cassandra folder back into web-local/target
8. Run locally using the `./start-migrator-role`
9. Write some deltas to Emo
10. Run cqlsh and peek at the tables. You should see your delta in both tables. They should be blocked in the new table
11. Kick off a migration using `curl -XPOST http://localhost:8082/migrator/1/migrate/<placement>/<migration_id> | jq .`
12. Unless you wrote an obscene amount of records originally (10's of millions), the migration process should finish within a few minutes
13. Checkout  [#116](https://github.com/bazaarvoice/emodb/pull/116)
14. Build and run emo locally
15. Witness that Emo is now running smoothly with blocked deltas

## Risk

### Level 

`High`

### Required Testing

`Regression` and `Manual`

### Risk Summary

The only immediate risk this PR poses is that compaction writes have been changed to CQL, which is significant. Additionally, it is also possible for the migration service to overload the C* ring, though I have implemented a throttling API just in case this occurs. 

In the future term, this PR poses a risk as the migrator service is being depended on to maintain consistency between tables. The best way to ensure that this is done successfully is probably to run Stash on both tables, and compare the results.

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
